### PR TITLE
Run cypher queries on web workers

### PR DESCRIPTION
**Pull request by irfannuri**
This PR enables running cypher queries on separate web workers (when available) so that web browser does not freeze and remains responsive. A brief list of facts about the implementation is listed below:

- There still exists an enduring connection in main thread to perform regular metadata updates and such needs
- When a cypher query is executed from editor (so `routedWriteTransaction` is called), a new web worker is created and the request is passed to that worker
- Each web worker creates a new driver object, session and connection which survive through request lifecycle
- Web worker and connection related resources are disposed either with the ordinary termination or cancellation of request.

With the use of web workers, it is possible to run multiple long running queries and still use the browser as seen in the attached screen shot. This was not possible, especially when the running query immediately started streaming results, which caused bolt driver to fully occupy cpu thus making UI non-responsive.

To disable web workers, for a limited time you can do so by `:config useCypherThread: false`.
This option will probably be removed in the future.

screen shot 2017-06-16 at 13 35 44

changelog: Use separate thread for running cypher queries

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -115,7 +116,8 @@
     "webpack": "^2.3.2",
     "webpack-bundle-analyzer": "^2.4.0",
     "webpack-dashboard": "^0.4.0",
-    "webpack-dev-server": "^2.4.2"
+    "webpack-dev-server": "^2.4.2",
+	"worker-loader": "^0.8.0"
```
Comment (oskarhane) [inline package.json]: Indentation

File: src/shared/services/bolt/bolt.js [hunk 2]
```diff
@@ -22,181 +22,102 @@ import { v4 } from 'uuid'
...
-    }
-    if (validateConnection === false) return resolve(driver)
-    _validateConnection(driver, resolve, reject)
-  })
-  return p
+function routedWriteTransaction (input, parameters, requestId = null, cancelable = false) {
+  if (enableWebWorkers && window.Worker) {
+    const id = requestId || v4()
+    /* eslint-disable import/no-webpack-loader-syntax */
+    const BoltWorkerModule = require('worker-loader!./boltWorker.js')
```
Comment (oskarhane) [inline src/shared/services/bolt/bolt.js]: Why not import as a regular module at the top of the file?

File: src/shared/services/bolt/bolt.js [hunk 3]
```diff
@@ -22,181 +22,102 @@ import { v4 } from 'uuid'
...
+import { runCypherMessage, cancelTransactionMessage, CYPHER_ERROR_MESSAGE, CYPHER_RESPONSE_MESSAGE, POST_CANCEL_TRANSACTION_MESSAGE } from './boltWorkerMessages'
 
-let _drivers = null
-let _useRoutingConfig = false
-let _routingAvailable = false
-const runningQueryRegister = {}
+let connectionProperties = null
+let boltWorkerRegister = {}
+let cancellationRegister = {}
+let enableWebWorkers = true
```
Comment (oskarhane) [inline src/shared/services/bolt/bolt.js]: Maybe we should start off by have this as a "experimental" setting in the config pane?
wdyt?

File: src/shared/services/bolt/boltMappings.js [hunk 4]
```diff
@@ -219,3 +219,31 @@ export const retrieveFormattedUpdateStatistics = (result) => {
 export const flattenProperties = (rows) => {
   return rows.map((row) => row.map((entry) => (entry && entry.properties) ? entry.properties : entry))
 }
+
+export const applyGraphTypes = (item) => {
```
Comment (oskarhane) [inline src/shared/services/bolt/boltMappings.js]: I'd like to see this function tested, so we don't introduce regressions if we need to make changes in it in the future.

File: src/shared/services/bolt/boltWorker.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    })
+  } else {
+    self.postMessage(cypherErrorMessage({
+      code: -1,
+      message: `Unknown message to Bolt Worker: ${messageType}`
+    }))
+  }
+}
+
+self.addEventListener('message', onmessage)
```
Comment (oskarhane) [inline src/shared/services/bolt/boltWorker.js]: What is self?

File: src/shared/services/bolt/boltMappings.js [hunk 6]
```diff
@@ -219,3 +219,31 @@ export const retrieveFormattedUpdateStatistics = (result) => {
...
+  if (Array.isArray(item)) {
+    return item.map(arrayItem => applyGraphTypes(arrayItem))
```
Comment (oskarhane) [inline src/shared/services/bolt/boltMappings.js]: Minor thing, but it's easier to read if you do this `item.map(applyGraphTypes)`.

File: src/shared/services/bolt/boltMappings.js [hunk 7]
```diff
@@ -219,3 +219,31 @@ export const retrieveFormattedUpdateStatistics = (result) => {
...
+  } else if (item.hasOwnProperty('labels') && item.hasOwnProperty('properties') && item.hasOwnProperty('identity')) {
+    return new neo4j.types.Node(neo4j.int(item.identity), item.labels, applyGraphTypes(item.properties))
+  } else if (item.hasOwnProperty('segments') && item.hasOwnProperty('start') && item.hasOwnProperty('end')) {
+    const start = applyGraphTypes(item.start)
+    const end = applyGraphTypes(item.end)
+    const segments = item.segments.map(segment => applyGraphTypes(segment))
```
Comment (oskarhane) [inline src/shared/services/bolt/boltMappings.js]: Same as above: `item.map(applyGraphTypes)`.

File: src/shared/services/bolt/boltMappings.test.js [hunk 8]
```diff
@@ -367,4 +368,221 @@ describe('boltMappings', () => {
...
+    test('should apply integer type', () => {
+      const rawNumber = { low: 5, high: 0 }
+      const typedNumber = applyGraphTypes(rawNumber)
+      expect(typedNumber instanceof neo4j.Integer).toEqual(true)
+    })
+
+    test('should apply node type', () => {
+      const rawNode = {
+        labels: ['Test'],
+        properties: [],
```
Comment (oskarhane) [inline src/shared/services/bolt/boltMappings.test.js]: Please add some properties of different types (integers, string, array, null, object) to make sure they come along.

File: src/shared/services/bolt/boltMappings.js [hunk 9]
```diff
@@ -219,3 +219,31 @@ export const retrieveFormattedUpdateStatistics = (result) => {
...
+  } else if (item.hasOwnProperty('relationship') && item.hasOwnProperty('start') && item.hasOwnProperty('end')) {
+    const start = applyGraphTypes(item.start)
+    const end = applyGraphTypes(item.end)
+    const relationship = applyGraphTypes(item.relationship)
+    return new neo4j.types.PathSegment(start, relationship, end)
+  } else if (item.hasOwnProperty('start') && item.hasOwnProperty('end') && item.hasOwnProperty('type')) {
+    return new neo4j.types.Relationship(neo4j.int(item.identity), neo4j.int(item.start), neo4j.int(item.end), item.type, applyGraphTypes(item.properties))
+  } else if (item.hasOwnProperty('low') && item.hasOwnProperty('high') && typeof item.low === 'number' && typeof item.high === 'number') {
+    return neo4j.int(item)
+  } else if (typeof item === 'object') {
```
Comment (oskarhane) [inline src/shared/services/bolt/boltMappings.js]: `null` is of type `object` so this will probably throw.

File: src/shared/services/bolt/boltMappings.test.js [hunk 10]
```diff
@@ -367,4 +368,221 @@ describe('boltMappings', () => {
...
+      expect(typedRelationships[0].identity instanceof neo4j.Integer).toEqual(true)
+      expect(typedRelationships[0].start instanceof neo4j.Integer).toEqual(true)
+      expect(typedRelationships[0].end instanceof neo4j.Integer).toEqual(true)
+      expect(typedRelationships[1] instanceof neo4j.types.Relationship).toEqual(true)
+      expect(typedRelationships[1].identity instanceof neo4j.Integer).toEqual(true)
+      expect(typedRelationships[1].start instanceof neo4j.Integer).toEqual(true)
+      expect(typedRelationships[1].end instanceof neo4j.Integer).toEqual(true)
+    })
+
+    test('should apply to custom object properties', () => {
```
Comment (oskarhane) [inline src/shared/services/bolt/boltMappings.test.js]: This test tests nodes again, not regular js objects. Please add tests for primitives as well, like `bool`, `null`, `string`, `float` etc.

File: src/shared/services/bolt/bolt.js (hunk 3 above)
Comment (irfannuri) [inline src/shared/services/bolt/bolt.js]: screen shot 2017-06-28 at 16 09 43

File: src/shared/services/bolt/boltMappings.js (hunk 4 above)
Comment (irfannuri) [inline src/shared/services/bolt/boltMappings.js]: Definitely. Done!

File: src/shared/services/bolt/boltMappings.js (hunk 6 above)
Comment (irfannuri) [inline src/shared/services/bolt/boltMappings.js]: Done

File: src/shared/services/bolt/boltMappings.js (hunk 7 above)
Comment (irfannuri) [inline src/shared/services/bolt/boltMappings.js]: done

File: src/shared/services/bolt/bolt.js (hunk 2 above)
Comment (irfannuri) [inline src/shared/services/bolt/bolt.js]: done

File: src/shared/services/bolt/boltWorker.js (hunk 5 above)
Comment (irfannuri) [inline src/shared/services/bolt/boltWorker.js]: It is needed by worker loader on the worker script

File: src/shared/services/bolt/boltMappings.test.js (hunk 8 above)
Comment (irfannuri) [inline src/shared/services/bolt/boltMappings.test.js]: done

File: src/shared/services/bolt/boltMappings.js (hunk 9 above)
Comment (irfannuri) [inline src/shared/services/bolt/boltMappings.js]: definitely! done.

**Comment by oskarhane** [comment]
I just rebased this over master and things are looking good. 
I removed the config UI and renamed the config option to `useCypherThread` (we might use service worker in the future, so no need to specify _what_ type of thread we use) and enabled it by default.

There's a jumpy behaviour going from spinner -> result that needs to be addressed before I want to merge this.
Also I want to do some more testing with more web browsers.

**Comment by oskarhane** [comment]
I've now addressed the "double frame insert animation" I was seeing.
I've also tested in major web browsers and had to Promise support in the worker file for it to work in IE 11.
To have the web worker working when the server serving the files is offline I had to make it an "inline worker". I've found no issues with it so far.

I'm approving this to be merged. Anyone else want to have a last check before merging @pe4cey @irfannuri @akollegger?

I'm merging after lunch tomorrow (CEST) unless someone tells me not to 🐎
